File: packages/api/cmd/sikuligrpc/examples/python/bootstrap.py
from __future__ import annotations

import os
import shutil
from pathlib import Path

from sikuligo.sikulix import _resolve_sikuli_binary


def ensure_sikuligo_on_path() -> str:
    existing = shutil.which("sikuligo")
    if existing:
        os.environ["SIKULIGO_BINARY_PATH"] = existing
        return existing

    source = Path(_resolve_sikuli_binary(None))
    install_dir = Path.cwd() / ".sikuligo" / "bin"
    install_dir.mkdir(parents=True, exist_ok=True)
    binary_name = "sikuligo.exe" if os.name == "nt" else "sikuligo"
    installed = install_dir / binary_name

    should_copy = True
    if installed.exists():
        should_copy = source.stat().st_size != installed.stat().st_size
    if should_copy:
        shutil.copy2(source, installed)
        if os.name != "nt":
            installed.chmod(0o755)

    current_path = os.environ.get("PATH", "")
    entries = [entry for entry in current_path.split(os.pathsep) if entry]
    if str(install_dir) not in entries:
        os.environ["PATH"] = f"{install_dir}{os.pathsep}{current_path}" if current_path else str(install_dir)

    os.environ["SIKULIGO_BINARY_PATH"] = str(installed)
    return str(installed)

File: packages/api/cmd/sikuligrpc/examples/node/bootstrap.mjs
import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { resolveSikuliBinary } from "@sikuligo/sikuligo";

function resolveOnPath(binaryName) {
  const cmd = process.platform === "win32" ? "where" : "which";
  const out = spawnSync(cmd, [binaryName], { encoding: "utf8" });
  if (out.status !== 0) {
    return "";
  }
  const first = String(out.stdout || "")
    .split(/\r?\n/)
    .map((line) => line.trim())
    .find((line) => line.length > 0);
  return first || "";
}

function ensureLocalInstallDirOnPath(installDir) {
  const current = process.env.PATH || "";
  const entries = current.split(path.delimiter).filter(Boolean);
  if (!entries.includes(installDir)) {
    process.env.PATH = `${installDir}${path.delimiter}${current}`;
  }
}

export function ensureSikuligoOnPath() {
  const binaryName = process.platform === "win32" ? "sikuligo.exe" : "sikuligo";
  const existing = resolveOnPath(binaryName);
  if (existing) {
    process.env.SIKULIGO_BINARY_PATH = existing;
    return existing;
  }

  const sourceBinary = resolveSikuliBinary();
  const installDir = path.resolve(process.cwd(), ".sikuligo", "bin");
  const targetBinary = path.join(installDir, binaryName);
  fs.mkdirSync(installDir, { recursive: true });

  let shouldCopy = true;
  if (fs.existsSync(targetBinary)) {
    const srcStat = fs.statSync(sourceBinary);
    const dstStat = fs.statSync(targetBinary);
    shouldCopy = srcStat.size !== dstStat.size;
  }
  if (shouldCopy) {
    fs.copyFileSync(sourceBinary, targetBinary);
    if (process.platform !== "win32") {
      fs.chmodSync(targetBinary, 0o755);
    }
  }

  ensureLocalInstallDirOnPath(installDir);
  process.env.SIKULIGO_BINARY_PATH = targetBinary;
  return targetBinary;
}
